internal/encryption/encryption.go

```diff
@@ -56,88 +62,327 @@ func (n NoOpEncryptor) DecryptJSON(input json.RawMessage, output any) error {
+
+	template, err := keyderivation.CreatePRFBasedKeyTemplate(prfKeyTemplate, aeadKeyTemplate)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create PRF based key template: %w", err)
+	}
+
+	handle, err := keyset.NewHandle(template)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create keyset handle: %w", err)
+	}
+
+	// Encrypt the keyset with the KEK AEAD.
+	buf := new(bytes.Buffer)
+	writer := keyset.NewBinaryWriter(buf)
+	err = handle.Write(writer, kekAEAD)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encrypt DEK: %w", err)
+	}
+	encryptedKeyset := buf.Bytes()
+
+	return &KmsEncryptor{
+		root:            *handle,
+		kekAEAD:         kekAEAD,
+		encryptedKeyset: encryptedKeyset,
+	}, nil
+}
+
+func NewKmsEncryptorWithKMS(uri string, v1client *awsv1kms.KMS, encryptedKeyset []byte) (*KmsEncryptor, error) {
+	kekAEAD, err := newClientWithAEAD(uri, v1client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create KMS client: %w", err)
+	}
+
+	reader := keyset.NewBinaryReader(bytes.NewReader(encryptedKeyset))
+	handle, err := keyset.Read(reader, kekAEAD)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read keyset: %w", err)
+	}
+
+	return &KmsEncryptor{
+		root:            *handle,
+		kekAEAD:         kekAEAD,
+		encryptedKeyset: encryptedKeyset,
+	}, nil
+}
+
+func (k *KmsEncryptor) GetEncryptedKeyset() []byte {
+	return k.encryptedKeyset
+}
+
+func deriveKeyset(root keyset.Handle, salt []byte) (*keyset.Handle, error) {
+	deriver, err := keyderivation.New(&root)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create deriver: %w", err)
+	}
+
+	derived, err := deriver.DeriveKeyset(salt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive keyset: %w", err)
+	}
+
+	return derived, nil
+}
+
+func (k *KmsEncryptor) Encrypt(subKey SubKey, cleartext []byte) ([]byte, error) {
+	encrypted, err := derivedEncrypt(k.root, subKey, cleartext)
+	if err != nil {
+		return nil, fmt.Errorf("failed to encrypt with derive: %w", err)
+	}
+
+	return encrypted, nil
+}
+
+func (k *KmsEncryptor) Decrypt(subKey SubKey, encrypted []byte) ([]byte, error) {
+	decrypted, err := derivedDecrypt(k.root, subKey, encrypted)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt with derive: %w", err)
+	}
+
+	return decrypted, nil
+}
+
+func derivedDecrypt(root keyset.Handle, subKey SubKey, encrypted []byte) ([]byte, error) {
+	derived, err := deriveKeyset(root, []byte(subKey))
+	if err != nil {
+		return nil, fmt.Errorf("failed to derive keyset: %w", err)
+	}
+
+	primitive, err := aead.New(derived)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create primitive: %w", err)
+	}
+
+	bytes, err := primitive.Decrypt(encrypted, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decrypt: %w", err)
+	}
+
+	return bytes, nil
+}
+
+func derivedEncrypt(root keyset.Handle, subKey SubKey, cleartext []byte) ([]byte, error) {
+	// TODO: Deriving might be expensive, consider caching the derived keyset.
```

Comment:
I think we should probably resolve this before it goes in (either we determine that this is cheap enough we don't need a cache, or we should add one).

Reply:
Yeah exactly. I suspect it is too heavy but will measure first.

Reply:
This is part one but I'll check it before the final PR.